.github/actions/bump-version/action.yml

```
@@ -0,0 +1,149 @@
+name: Bump Version
+author: Tommy Gatti
+description: Action for bumping a version string
+inputs:
+  version:
+    required: true
+    type: string
+    description: The version string that will be bumped
+  versioning-scheme:
+    required: true
+    type: string
+    description: |
+      The type of versioning scheme used. Currently supports 'calver-minor', 'semver-major-minor', where:
+      'calver-minor': 'YYYY.OM.MINOR'.
+      'semver-major-minor': 'MAJOR.MINOR'.
+  bump-type:
+    required: true
+    type: string
+    description: |
+      The type of bump. This changes depending on the `versioning-scheme`.
+      'calver-minor': 'major' or 'minor'.
+      'semver-major-minor': 'major' or 'minor'.
+outputs:
+  before:
+    description: Version before being bumped. Equivalent to the input `version`.
+    value: ${{ steps.results.outputs.before }}
+  after:
+    description: Version after being bumped.
+    value: ${{ steps.results.outputs.after }}
+  type:
+    description: Type of bump. Equivalent to the input `bump-type`.
+    value: ${{ steps.results.outputs.type }}
+runs:
+  using: composite
+  steps:
+    - name: Check Versioning Scheme Validity
+      shell: bash
+      env:
+        VALID_VERSIONING_SCHEMES: calver-minor semver-major-minor
+
+      run: |
+        # check validity of versioning schemes
+        found=false
+        for valid_scheme in ${{ env.VALID_VERSIONING_SCHEMES }}; do
+          if [[ "${{ inputs.versioning-scheme }}" == "${valid_scheme}" ]]; then
+            found=true
+            break
+          fi
+        done
+        if [[ "${found}" == "false" ]]; then
+          echo "::error::${{ inputs.versioning-scheme }} is not supported by this action. Please check the list of allowed 'versioning-scheme's."
+          exit 1
+        fi
+
+    - name: Check Bump Type Validity for ${{ inputs.versioning.scheme }}
+      shell: bash
+      env:
+        VALID_CALVER_MINOR_BUMP_TYPES: major minor
+        VALID_SEMVER_MAJOR_MINOR_BUMP_TYPES: major minor
+      run: |
+        # check validity of bump types for each versioning scheme
+        found=false
+
+        if [[ "${{ inputs.versioning-scheme }}" == "calver-minor" ]]; then
+          for valid_bump in ${{ env.VALID_CALVER_MINOR_BUMP_TYPES }}; do
+            if [[ "${{ inputs.bump-type }}" == "${valid_bump}" ]]; then
+              exit 0
+            fi
+          done
+        elif [[ "${{ inputs.versioning-scheme }}" == "semver-major-minor" ]]; then
+          for valid_bump in ${{ env.VALID_SEMVER_MAJOR_MINOR_BUMP_TYPES }}; do
+            if [[ "${{ inputs.bump-type }}" == "${valid_bump}" ]]; then
+              exit 0
+            fi
+          done
+        fi
```

```suggestion
        declare -A valid_bumps
        for scheme in "${{VALID_VERSIONING_SCHEMES}}"; do
          scheme_env_var = $(echo "VALID_${VAR_1^^}_TYPES" | tr '-' '_')
          valid_bumps[${scheme}] = ${{ $scheme_env_var }}
        done
        for valid_bump in ${!valid_bumps[@]}; do
          if [[ "${{ inputs.bump-type }}" == "${valid_bump}" ]]; then
            exit 0
          fi
        done
```
This makes a bash associative array out of valid bump types for each scheme.

It is of dubious benefit in this case because there are only two schemes, but if that is ever increased then the code doesn't need to change, the new scheme and bump types just need to be added to the environment variables.

It does require `VALID_VERSIONING_SCHEMES` to be scoped globally.

Also I have not tested this code except for the mess that is `$(echo "VALID_${VAR_1^^}_TYPES" | tr '-' '_')`. That could be a lot simpler if the variable names were chosen to fit this idea.

---

I might need you to run through this with me :D

---

Updated this section now - figured out a way to do it without creating an associative array, as wel only ever need one of those keys in the array at any one time, so we don't need to create the entire thing.